src/v/cluster/members_table.cc

```diff
@@ -140,6 +142,56 @@ std::vector<model::node_id> members_table::get_decommissioned() const {
     return ret;
 }
 
+std::error_code
+members_table::apply(model::offset version, maintenance_mode_cmd cmd) {
+    _version = model::revision_id(version());
+
+    const auto target = _brokers.find(cmd.key);
+    if (target == _brokers.end()) {
+        return errc::node_does_not_exists;
+    }
+
+    // no rules to enforce when disabling maintenance mode
+    const auto enable = cmd.value;
+    if (!enable) {
+        target->second->set_maintenance_state(
+          model::maintenance_state::inactive);
+        return errc::success;
+    }
+
+    if (
+      target->second->get_maintenance_state()
+      == model::maintenance_state::active) {
+        return errc::success;
+    }
+
+    /*
+     * enforce one-node-at-a-time in maintenance mode rule
```

Comment:
my mental model for the controller log is that its a classic replicated state machine in which the result of a command isn't known until replay. in this view, yes, incompatible commands can be in log and the error we are generating here is ultimately returned back to the client making the request. I don't think this would stop the replay of other message, but perhaps I'm missing something there. can you say more about that concern?

regarding validation up front: (1) at a minimum we could / should do validation up front to avoid unnecessary noise in the log, but (2) more than a naive check of in-memory state would be needed to maintain serializability. I think @rystsov has demonstrated this in some other area where we propagate the expected term down into the storage layer. I'm less clear how this mode of upfront validation and strong consistency work.

I'll go a head add the basic up front validation, tho, I think that makes a lot of sense.

Reply:
update on this I see we aren't doing any other validation in the members frontend so presumably we'd need to bring in a reference to the members table. what do you think @mmaslankaprv ?

Reply:
i think basic validation (without using `term`) and linearizable barriers is just fine in the topics frontend. That would be an optimization. Final validation will be made during applying entries to STM.

Reply:
> i think basic validation (without using term) and linearizable barriers is just fine in the topics frontend. That would be an optimization. Final validation will be made during applying entries to STM.

are you saying the validation should be in topics frontend and not in the members_frontend? do you have an example pointer of a similar type of validation being done?

Reply:
@mmaslankaprv just to clarify, are you saying that creating topics is related to maintenance mode?

Reply:
Spoke with Michal offline. The current version in the PR is the easiest solution that provides strong consistency. Other validation can be done upfront with some more effort to avoid writing commands that are known to fail but are optimizations.

Will add this optimization along with the deduplication to avoid flapping of the service in a follow up PR.